service-base/src/main/java/org/eclipse/hono/service/registration/BaseRegistrationService.java

```diff
@@ -124,28 +133,34 @@ protected String getEventBusAddress() {
         final String gatewayId = request.getGatewayId();
         final SpanContext spanContext = request.getSpanContext();
 
+        final Span span = newChildSpan(SPAN_NAME_ASSERT_DEVICE_REGISTRATION, spanContext, tenantId, deviceId, gatewayId);
         if (tenantId == null || deviceId == null) {
+            TracingHelper.logError(span, "missing tenant and/or device");
+            span.finish();
             return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST));
-        } else if (gatewayId == null) {
-            log.debug("asserting registration of device [{}] with tenant [{}]", deviceId, tenantId);
-            final Future<RegistrationResult> result = Future.future();
-            assertRegistration(tenantId, deviceId, spanContext, result.completer());
-            return result.map(res -> {
-                return request.getResponse(res.getStatus())
-                        .setDeviceId(deviceId)
-                        .setJsonPayload(res.getPayload())
-                        .setCacheDirective(res.getCacheDirective());
-            });
         } else {
-            log.debug("asserting registration of device [{}] with tenant [{}] for gateway [{}]",
-                    deviceId, tenantId, gatewayId);
             final Future<RegistrationResult> result = Future.future();
-            assertRegistration(tenantId, deviceId, gatewayId, spanContext, result.completer());
+            if (gatewayId == null) {
+                log.debug("asserting registration of device [{}] with tenant [{}]", deviceId, tenantId);
+                assertRegistration(tenantId, deviceId, span, result.completer());
```

Comment:
Yes, I'm aware of that. Just thought that using `result.completer()` makes it more explicit that a Handler is used here - also because other places in Hono still use that as well.